exercises/react/example.cr

```diff
@@ -0,0 +1,80 @@
+alias CellValue = Int32
+
+private class Cell
+  @dependents = [] of ComputeCell
+
+  def initialize(initial : CellValue)
+    @value = initial
+  end
+
+  getter :value
+
+  protected getter :dependents
+  protected def propagate; end
+  protected def fire_callbacks; end
+end
+
+private class InputCell < Cell
+  def value=(new_value : CellValue)
+    @value = new_value
+    @dependents.each(&.propagate)
+    @dependents.each(&.fire_callbacks)
+  end
+end
+
+private class ComputeCell < Cell
+  alias CallbackID = Int32
+
+  @callbacks_given = 0
+  @callbacks = {} of CallbackID => CellValue ->
+  @last_callback_value : CellValue
+
+  def initialize(new_value : -> CellValue, *inputs : Cell)
+    super(new_value.call)
+    @new_value = new_value
+    @last_callback_value = @value
+    inputs.each { |i| i.dependents << self }
+  end
+
+  def add_callback(&block : CellValue ->) : CallbackID
+    @callbacks_given += 1
+    @callbacks[@callbacks_given] = block
+    @callbacks_given
+  end
+
+  def remove_callback(id : CallbackID)
+    @callbacks.delete(id)
+  end
+
+  protected def propagate
+    v = @new_value.call
+    return if v == @value
+    @value = v
+    @dependents.each(&.propagate)
+  end
+
+  protected def fire_callbacks
+    return if @value == @last_callback_value
+    @last_callback_value = @value
+    @callbacks.each_value { |c| c.call(@value) }
+    @dependents.each(&.fire_callbacks)
+  end
+end
+
+class Reactor
+  # Note that the Reactor does not keep any extra state *in this example solution*.
+  # In other solutions, it's possible that Reactors may keep track of their cells,
+  # take reponsibility for value propagation and callback firing, etc.
+
+  def create_input(initial : CellValue) : InputCell
+    InputCell.new(initial)
+  end
+
+  def create_compute(input : Cell, &block : CellValue -> CellValue) : ComputeCell
+    ComputeCell.new(->{ block.call(input.value) }, input)
+  end
+
+  def create_compute(input1 : Cell, input2 : Cell, &block : (CellValue, CellValue) -> CellValue) : ComputeCell
```

Comment:
I wished that I could create a general function of the form create_compute(*inputs : Cell, &block : Something -> CellValue) : ComputeCell.

However, I was not able to do such a thing. I think the general problem is you cannot write a type for "A function that takes as many inputs as there are values in inputs". Therefore, there is nothing I can fill in for Something, above.

Now, you could do def create_compute(*inputs : Cell, &block : Array(CellValue) -> CellValue) : ComputeCell (and the below line would be something like ComputeCell.new(inputs.map(&.value).to_a}, input1, input2). However, now note that we have lost the ability to check that the types match. Someone could create a compute cell with 1 input but whose compute function takes 2 inputs, or vice versa.

So I did it this way to maintain that the compiler can check that the arities match up. Obviously the disadvantage is if we want compute cells that take 3 inputs we have to write a third function, etc.

If you think there is a better way to deal with the problem, let me know.